service/topology_coordinator.cc

```diff
@@ -67,6 +68,652 @@ future<inet_address> wait_for_ip(raft::server_id id, const raft_address_map& am,
+        if (can_accept_load(mig_streaming_info)) {
+            rtlogger.info("upsize_tablet_replicas_to_rf: accepted migration {} from {} to {}", source_tablet, src, dst);
+            apply_load(mig_streaming_info);
+            stats.for_dc(dc).migrations_produced++;
+            new_replicas.emplace_back(dst);
+        } else {
+            rtlogger.info("upsize_tablet_replicas_to_rf: skipping migration {} from {} to {} because of streaming load",
+                source_tablet, src, dst);
+            // Shards are overloaded with streaming. Do not include the migration in the plan, but
+            // continue as if it was in the hope that we will find a migration which can be executed without
+            // violating the load. Next make_plan() invocation will notice that the migration was not executed.
+            // We should not just stop here because that can lead to underutilization of the cluster.
+            // Just because the next migration is blocked doesn't mean we could not proceed with migrations
+            // for other shards which are produced by the planner subsequently.
+            skipped_migrations++;
+            stats.for_dc(dc).migrations_skipped++;
+            if (skipped_migrations >= max_skipped_migrations) {
+                stats.for_dc(dc).stop_skip_limit++;
+                break;
+            }
+        }
+
+        target_info.tablet_count += 1;
+        target_info.update();
+
+        src_shard_info.tablet_count -= 1;
+        if (src_shard_info.tablet_count == 0) {
+            push_back_shard_candidate.cancel();
+            src_node_info.shards_by_load.pop_back();
+        }
+
+        src_node_info.tablet_count -= 1;
+        src_node_info.update();
+        if (src_node_info.tablet_count == 0) {
+            push_back_node_candidate.cancel();
+            nodes_by_load.pop_back();
+        }
+    }
+
+    co_await utils::clear_gently(nodes);
+
+    std::copy(other_dc_replicas.begin(), other_dc_replicas.end(), std::back_inserter(new_replicas));
+
+    co_return upsize_result{new_replicas, tablet_reallocation_status::success};
+}
+
+future<tablet_reallocation_result>
+reallocate_tablets_for_new_rf(schema_ptr s, locator::token_metadata_ptr tm,
+    std::unordered_map<sstring, sstring> const& ks_options) {
+
+    auto const& dcs = tm->get_topology().get_datacenters();
+    std::unordered_map<sstring, size_t> new_rf_per_dc;
+    for (auto const& [option, value] : ks_options) {
+        if (option == "replication_factor")
+            continue;
+        new_rf_per_dc[option] = std::stoul(value);
+    }
+    auto rf_it = ks_options.find("replication_factor");
+    if (rf_it != ks_options.cend()) {
+        auto rf = std::stoul(rf_it->second);
+        for (auto const& dc : dcs) {
+            if (!new_rf_per_dc.contains(dc)) {
+                new_rf_per_dc[dc] = rf;
+            }
+        }
+    }
+
+    rtlogger.info("reallocate_tablets_for_new_rf: computed new_rf_per_dc: {}", new_rf_per_dc);
```

Comment:
nit: Probably not "computed", but provided? These are the numbers that came directly from ALTER statement